Improve <input type=color alpha> performance on macOS

annevk opened the pull request:
#### c6709039985b45fbabb8cf2256cb1d1b00baf724
<pre>
Improve &lt;input type=color alpha&gt; performance on macOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=286803">https://bugs.webkit.org/show_bug.cgi?id=286803</a>
<a href="https://rdar.apple.com/143955048">rdar://143955048</a>

Reviewed by Aditya Keerthi.

Using data: URLs here resulted in annoying flickering when moving the
alpha slider quickly.

Using SVG directly in the shadow tree ran into
<a href="https://bugs.webkit.org/show_bug.cgi?id=287472">https://bugs.webkit.org/show_bug.cgi?id=287472</a> so therefore we use a
div element with a polygon clip-path. This results in additional render
layers, but websites will likely not have a lot of these controls so
that seems okay.

* LayoutTests/fast/forms/color/color-input-swatch-expected.txt:
* LayoutTests/fast/forms/color/color-input-swatch.html:
* LayoutTests/platform/mac/fast/forms/color/color-input-swatch-expected.txt:
* LayoutTests/platform/mac/fast/forms/color/input-appearance-color-expected.txt:
* Source/WebCore/html/ColorInputType.cpp:
(WebCore::ColorInputType::createShadowSubtree):
(WebCore::ColorInputType::updateColorSwatch):
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::createColorWellSwatchSubtree):
* Source/WebCore/rendering/mac/RenderThemeMac.h:
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
(WebCore::RenderThemeMac::createColorWellSwatchSubtree):
(WebCore::RenderThemeMac::setColorWellSwatchBackground):

Canonical link: <a href="https://commits.webkit.org/290228@main">https://commits.webkit.org/290228@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cebe08f2d848044f08afce00f85b55ba23ced45e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89293 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8817 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44129 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94278 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40054 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91344 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9205 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/17142 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68799 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26467 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92295 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7062 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81041 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49160 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6808 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/35426 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39161 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77164 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36411 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96108 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16473 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12095 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77673 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16729 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76829 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76972 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19000 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21388 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19989 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9602 "Built successfully") | 
| [  ~~🛠 🧪 unsafe-merge~~](https://ews-build.webkit.org/#/builders/22/builds/13994 "The change is no longer eligible for processing.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16487 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21798 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16228 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19679 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18009 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on LayoutTests/fast/forms/color/color-input-swatch-expected.txt:
```diff
@@ -1,102 +1,85 @@
 getComputedStyle(swatch).backgroundColor is rgb(0, 0, 0)
-getComputedStyle(swatch).backgroundImage is none
-getComputedStyle(swatch).backgroundSize is auto
+swatch.innerHTML is
```
This doesn't look quite right. Maybe you want a `shouldBe`-like function here to clearly state that empty string is the expectation? Other logged values may benefit from `shouldBe`, too.

annevk replied:
I don't really want to state the expectations as they can differ per platform, as they do here.

aproskuryakov replied:
Even this one, empty `innerHTML`?

That said, color components differing per platform (and I'm sure also per configuration) is only more of a reason to have expectations. Possibly, saying that the numbers should be within a certain range. Currently, you are asking everyone who sees the test fail to investigate for themselves whether the different number means failing or passing, and that's rather hard.

annevk replied:
They only differ between macOS and everywhere else at this point. And only macOS has a deeper subtree. Anyone touching the shadow tree for this component would know why this test regressed.

aproskuryakov replied:
I hear your explanations, but I still really dislike tests that just dump results, as existing ones that do so are a constant nuisance.

annevk committed:
Improve <input type=color alpha> performance on macOS
https://bugs.webkit.org/show_bug.cgi?id=286803
rdar://143955048

Reviewed by Aditya Keerthi.

Using data: URLs here resulted in annoying flickering when moving the
alpha slider quickly.

Using SVG directly in the shadow tree ran into
https://bugs.webkit.org/show_bug.cgi?id=287472 so therefore we use a
div element with a polygon clip-path. This results in additional render
layers, but websites will likely not have a lot of these controls so
that seems okay.

* LayoutTests/fast/forms/color/color-input-swatch-expected.txt:
* LayoutTests/fast/forms/color/color-input-swatch.html:
* LayoutTests/platform/mac/fast/forms/color/color-input-swatch-expected.txt:
* LayoutTests/platform/mac/fast/forms/color/input-appearance-color-expected.txt:
* Source/WebCore/html/ColorInputType.cpp:
(WebCore::ColorInputType::createShadowSubtree):
(WebCore::ColorInputType::updateColorSwatch):
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::createColorWellSwatchSubtree):
* Source/WebCore/rendering/mac/RenderThemeMac.h:
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
(WebCore::RenderThemeMac::createColorWellSwatchSubtree):
(WebCore::RenderThemeMac::setColorWellSwatchBackground):

Canonical link: https://commits.webkit.org/290228@main
